frame/staking-rewards/src/lib.rs

```diff
@@ -780,9 +773,102 @@ pub mod pallet {
 			});
 			Ok([*position, new_position])
 		}
+
+		#[transactional]
+		fn claim(who: &Self::AccountId, position: &Self::PositionId) -> DispatchResult {
+			let keep_alive = false;
+
+			let mut stake =
+				Stakes::<T>::try_get(position).map_err(|_| Error::<T>::StakeNotFound)?;
+			ensure!(who == &stake.owner, Error::<T>::OnlyStakeOwnerCanUnstake);
+
+			let pool_id = stake.reward_pool_id;
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let early_unlock = stake.lock.started_at.safe_add(&stake.lock.duration)?
+				>= T::UnixTime::now().as_secs();
+
+			(rewards_pool, stake) =
+				Self::collect_rewards(&pool_id, rewards_pool, stake, early_unlock, keep_alive)?;
+
+			RewardPools::<T>::insert(pool_id, rewards_pool);
+			Stakes::<T>::insert(position, stake);
+
+			Self::deposit_event(Event::<T>::Claimed { owner: who.clone(), position_id: *position });
+
+			Ok(())
+		}
 	}
 
 	impl<T: Config> Pallet<T> {
+		/// Transfers the rewards a staker has earned while updating the provided `rewards_pool`.
+		///
+		/// # Params
+		/// * `pool_id` - Pool identifier
+		/// * `mut rewards_pool` - Rewards pool to update
+		/// * `stake` - Stake position
+		/// * `early_unlock` - If there should be an early unlock penalty
+		/// * `keep_alive` - If the transaction should be kept alive
+		pub(crate) fn collect_rewards(
+			pool_id: &T::RewardPoolId,
+			mut rewards_pool: RewardPoolOf<T>,
+			mut stake: StakeOf<T>,
+			early_unlock: bool,
+			keep_alive: bool,
+		) -> Result<(RewardPoolOf<T>, StakeOf<T>), DispatchError> {
+			let mut inner_rewards = rewards_pool.rewards.into_inner();
+
+			for (asset_id, reward) in inner_rewards.iter_mut() {
+				let inflation = stake.reductions.get(asset_id).cloned().unwrap_or_else(Zero::zero);
+				let claim = if rewards_pool.total_shares == Zero::zero() {
+					Zero::zero()
+				} else {
+					reward
+						.total_rewards
+						.safe_mul(&stake.share)?
+						.safe_div(&rewards_pool.total_shares)?
+						.safe_sub(&inflation)?
+				};
+				let claim = if early_unlock {
+					(Perbill::one() - stake.lock.unlock_penalty).mul_ceil(claim)
+				} else {
+					claim
+				};
+				let claim = sp_std::cmp::min(
+					claim,
+					reward.total_rewards.safe_sub(&reward.claimed_rewards)?,
+				);
+
+				reward.claimed_rewards = reward.claimed_rewards.safe_add(&claim)?;
+
+				// REVIEW: There has to be a better way to do this
```

Comment:
This is correct, and should be kept as it is :) 
Why you think there is a better way?

Reply:
Felt rather verbose initially. However, this is the cleaned-up version of it.
a17b58323c20394159d31ada738e76134740fcdf